### ISSUE TITLE
Blocking code is not allowed in sync #[behaviour(event_process = true)] trait implementation

### ISSUE THREAD

**Issue by Frederik-Baetens** [libp2p/rust-libp2p#2474]
As the traits generated by event_process=true are standard sync traits, there's no warning to the user that blocking code is not allowed.

I realize that async fn in traits isn't possible as of yet, but in the meantime it seems wise to document this, or alternatively devise a mechanism whereby these sync method calls are spawned on a separate thread, or with block_on.

Something like this already exists today:

```
        libp2p::swarm::SwarmBuilder::new(transport, mybehaviour, peer_id)
            .executor(Box::new(|fut| {
                tokio::spawn(fut);
            }))
            .build()
```

Perhaps we could allow the user to pass an additional closure, for blocking calls, like so:

```
        libp2p::swarm::SwarmBuilder::new(transport, mybehaviour, peer_id)
            .executor(Box::new(|fut| {
                tokio::spawn(fut);
            }))
            .blocking_spawn(Box::new(|function| {
                tokio::task::spawn_blocking(function);
            }))
            .build()
```
(Idk if my closure is correct)

Alternatively we could add feature flags for various executors, which would remove the need to pass these kinds of closures (I think, correct me if that's wrong).

Would it be possible to use [the async trait crate](https://crates.io/crates/async-trait) for this?

Attached you can find a demonstration of problems that can occur when using blocking code in the trait implementation:

https://user-images.githubusercontent.com/18269095/152414077-06be9795-fdbc-4e3e-8983-6f7b12808b8c.mp4

**Comment by rkuhn** [comment]
In general I find that swarm event handling should only be used for high-level coordination tasks — I’m currently reimplementing our (so far private) libp2p-streaming-response crate to avoid `Swarm::poll` for sending response messages. Instead, it makes more sense to directly interact with substreams from other app code, bypassing the swarm and communicating directly with the per-connection task spawned using the Network’s `.executor`.

I agree this should be documented and can also help with that.

Unfortunately there is no magic fix we can propose, since the purpose of NetworkBehaviourEventProcess is to provide an exclusive reference to the behaviour — spawning the code on a separate thread would not solve the blocking problem.

**Comment by mxinden** [comment]
As an aside, in the long run I would like to remove ` NetworkBehaviourEventProcess` and `event_process=true` entirely in favor of `event_process=false`. As far as I can tell there is no use-case one could not fulfill with the latter. Please tell me in case I am missing one. This is likely a discussion for a new GitHub issue.

What is the reason you are not using `event_process=false`? You could then spawn your blocking logic outside of libp2p, e.g. in whatever component is polling the swarm.

---

> libp2p-streaming-response crate to avoid Swarm::poll for sending response messages. Instead, it makes more sense to directly interact with substreams from other app code, bypassing the swarm and communicating directly with the per-connection task spawned using the Network’s .executor.

I think rust-libp2p should offer something along the lines of https://github.com/libp2p/rust-libp2p/pull/1947/ to support this use-case.

**Comment by rkuhn** [comment]
The PR you reference contains the code that I am currently replacing at Actyx :-) I found that negotiating a new substream for each message incurs unacceptable overhead. I’ll open a new PR once the work is finished and I’m satisfied with the quality, which may still take 2–3 weekends.

Regarding the removal of `event_process = true`: that would make it impossible to write encapsulated behaviours that compose other behaviours into nice building blocks, at least with the current infrastructure. The “umbrella” behaviour will need to have access to the events generated by its sub-behaviours.

**Comment by Frederik-Baetens** [comment]
> Regarding the removal of event_process = true: that would make it impossible to write encapsulated behaviours that compose other behaviours into nice building blocks

This is the main reason why I'm using event_process = true, it's possible to access the other behaviours trough the swarm, but I dislike having all my event handling code in the main swarm loop, and once you start encapsulating the behaviours into separate functions it gets a bit more difficult. I don't know if it's possible/easy/clean to pass the swarm into the handling functions. Event_process = true provides me with a clean structure for my code, and I wouldn't know how to as cleanly structure my code otherwise.

In fact, I'd prefer if I didn't have to poll the swarm in a loop at all. I think that's really weird. I assume it's to be able to stay agnostic towards the async executor, but I'd prefer if I just had to implement some traits, which then get called automatically after I just call some .run() method that libp2p provides. This workflow would be similar to a classic webserver, for example in axum or rocket, I just write some functions, tell axum/rocket which routes they belong to, and call .run(). I don't continuously need to poll axum/rocket myself in a loop.

Also, I've had some issues trying to use event_process=false without a custom outevent, which I found a bit messy. The docs make it seem like there's a way to use a custom behaviour with event_process=false without a custom outevent. Is there an example showing this?

Correct me if I'm wrong on any of this. I'd love to find a better structure for my code.

**Comment by mxinden** [comment]
> I found that negotiating a new substream for each message incurs unacceptable overhead.

How come? Due to message size? Negotiation should be 0-RTT with [`V1Lazy`](https://docs.rs/multistream-select/latest/multistream_select/enum.Version.html#variant.V1Lazy). And potentially even better far far in the future with https://github.com/libp2p/specs/pull/349.

**Comment by rkuhn** [comment]
Hmm, when I read those V1Lazy docs I concluded that it has too many edges to be used as a general purpose tool. For example, we’re using libp2p specifically because we want dynamic protocol negotiation, i.e. neither side knows the precise choice of protocol when a request is initiated. Otherwise I could just use TCP and put a small muxer in front.

On the conceptual level, I find it valuable to identify a response stream as a single entity on the underlying transport. This will allow solving fairness or quota issues in the future.

Another concern is operational: we need debug logging to be useful, which excludes per-message lines — per-stream is necessary, though.

In summary, messages are a necessary primitive, and streams of messages are another but very different necessary primitive. Please don’t remove them.

**Comment by rkuhn** [comment]
I forgot one very important point: with separate substreams per message it is impossible to implement back-pressure.

**Comment by mxinden** [comment]
> Also, I've had some issues trying to use event_process=false without a custom outevent, which I found a bit messy. The docs make it seem like there's a way to use a custom behaviour with event_process=false without a custom outevent. Is there an example showing this?

That is not possible. Would you mind sending in a patch fixing the docs?

**Comment by mxinden** [comment]
> I forgot one very important point: with separate substreams per message it is impossible to implement back-pressure.

Yes this is a good point. I have a couple of thoughts on this, especially in combination with protocols like bitswap. Though this is likely for another issue.

(As a side note, if I am not mistaken, QUIC does have backpressure on channel creation. Though one would need to enforce it within `ProtocolsHandler`.)
